akvo/rsr/usecases/utils.py

```diff
@@ -0,0 +1,125 @@
+    common_ancestor = get_first_common_item(lhs_lineage, rhs_lineage)
+    # Remove all projects that are ancestors of the nearest common ancestor and make the nearest common ancestor
+    # as the root of the hierarchy.
+    return set(lhs_lineage).symmetric_difference(set(rhs_lineage)) | {common_ancestor}
+
+
+@dataclass(frozen=True)
+class TreeNode:
+    item: Any
+    children: List['TreeNode'] = field(default_factory=list)
+
+
+def make_tree_from_list(items, parent_attr) -> List[TreeNode]:
+    tree = []
+    ids = [it['id'] for it in items]
+    lookup = {item['id']: TreeNode(item=item) for item in items}
+    for item in items:
+        item_id = item['id']
+        node = lookup[item_id]
+        parent_id = item.get(parent_attr, None)
+        # Root node, or any items with parents outside of the project ancestry are added to the tree
+        if not parent_id or parent_id not in ids:
+            tree.append(node)
+        else:
+            lookup[parent_id].children.append(node)
+    return tree
+
+
+def filter_trees(trees: List[TreeNode], predicate: Callable[[TreeNode], bool]) -> List[TreeNode]:
+    return [tree for tree in trees if is_tree_contains(tree, predicate)]
+
+
+def is_tree_contains(tree: TreeNode, predicate: Callable[[TreeNode], bool]) -> bool:
+    if predicate(tree):
+        return True
+    for child in tree.children:
+        if is_tree_contains(child, predicate):
+            return True
+    return False
+
+
+def get_leaf_item(node: TreeNode):
+    return get_leaf_item(node.children[0]) if len(node.children) else node.item
+
+
+def make_target_parent_map(trees: Iterable[TreeNode], project_attr, original_project_id, target_project_id) -> Dict[int, int]:
+    target_map = {}
+    for node in trees:
+        if len(node.children) == 0:
+            target_map[node.item['id']] = None
+            continue
+        first_leaf = get_leaf_item(node.children[0])
+        if len(node.children) == 1 \
+                and node.item[project_attr] == original_project_id \
+                and first_leaf[project_attr] == target_project_id:
+            target_map[node.item['id']] = first_leaf['id']
+            continue
+        if len(node.children) == 1 \
+                and first_leaf[project_attr] == original_project_id:
+            target_map[first_leaf['id']] = None
+            continue
+        second_leaf = get_leaf_item(node.children[1])
+        if len(node.children) == 2 \
+                and first_leaf[project_attr] == original_project_id:
+            target_map[first_leaf['id']] = second_leaf['id'] if second_leaf[project_attr] == target_project_id else None
+            continue
+        if len(node.children) == 2 \
+                and second_leaf[project_attr] == original_project_id:
+            target_map[second_leaf['id']] = first_leaf['id'] if first_leaf[project_attr] == target_project_id else None
```

Comment:
Why are only max 2 children considered?

It would be good to have comments on what this method is supposed to be doing and what the conditions mean.

I suppose it finds the IDs of of objects in the target project (new parent project?) that will act as the new parents of the node item?